### PR TITLE
Types

### LINKED ISSUE #27: Segment types
As discussed earlier, there are multiple segment types besides sponsor spots. This issue tracks the API changes necessary to implement them.

There are many things that would ideally be changed, but also new features are implementable without those changes. The minimal set of changes is highlighted with **Minimal change**.

All renamings imply the creation of new API endpoints without removal of old ones methods

# Database
- **NO** Rename table `sponsorTimes` into `segmentTimes`
- [ ] **Minimal change** Add `type` ENUM column with values (to be discussed):
  - `null` -- type unknown and for backwards compatibility
  - `"intro"` -- for vide intros, usually appear before channel logo (not all channels have this)
  - `"sponsor"` -- for currently tracked sponsors
  - `"merch"` -- channel's custom merch (not an outside sponsor)
  - `"social"` -- We are on Twitter/FaceBook/etc.
  - `"buttons"` -- Comment!/Like!/Subscribe!
  - `"patreon"` -- Only for patreon.com announcements that do not contain 

Some of the categories might overlap (`"sponsor"` and `"merch"`; `"social"` and `"buttons"` and `"patreon"`), but this is intentional and enables finer granularity.

# API
- [ ] **Minimal change** To current `GET /api/postVideoSponsorTimes` add `type` which is a string (one of recognized types).
- [ ] Create `GET /api/videoSegmentTimes` to be used instead of `GET /api/getVideoSponsorTimes`
  - [ ] It sends the same data, but with `type`.
- [ ] Create `POST /api/videoSegmentTimes` to be used instead of `GET /api/postVideoSponsorTimes`
  - [ ] Use actual JSON instead of URL parameters, allow submitting multiple segments (for the same video). No mechanism for segment deletion yet. Schema:
    - `userID` -- same as now
    - `videoID` -- string, same as now
    - `segments` -- array of Objects like so:
       - `startTime` -- float, in seconds
       - `endTime` -- float, in seconds
       - `type` -- string, one of known types
- [ ] Update `GET /api/postVideoSponsorTimes` to insert `type = "sponsor"`
- [ ] Ensure `GET /api/getVideoSponsorTimes` returns only segments with `type = "sponsor"`

# NOTE: BEFORE MERGING:
You need to update the table to insert the `type` column. Since SQLLite does not have `ENUM`s, it can be `TEXT` type.
```
ALTER TABLE sponsorTimes ADD "type" TEXT;
UPDATE sponsorTimes SET "type"="sponsor";
```

### PR DESCRIPTION
**Pull request by bershanskiy**
Fixes #27.

I finally got around to pushing this PR.

I intentionally didn't do pretty much any refactoring to make it easier to merge.

I couldn't resist making the first three commits simply because of how trivial they are yet how much they improve convenience.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -1,22 +1,23 @@
...
 
 //hashing service
 var crypto = require('crypto');
 
+let config = JSON.parse(fs.readFileSync('config.json'));
+
+// Default to production mode (for development mode set
+// development mode set it to 'development')
+// do not return full stacktrace to random people on internet
+//app.set('env', config.mode || 'production');
```
Comment (ajayyy) [inline index.js]: Should this be commented out?

Comment (bershanskiy) [reply index.js]: I wasn't sure whether you wanted this, so I commented it out for compatibility purposes. Ideally, yes Express would run in production mode on production.

**Comment by ajayyy** [comment]
Ah, so the client will handle the types. One small issue with this plan is: What should happen if two similar sponsors have different types?

**Comment by bershanskiy** [comment]
> Ah, so the client will handle the types.

Yes. The server has to send all types to prevent the server from knowing what the user's preferences of skipping are (e.g., if a user wants to skip only sponsors, social stuff and channel logos, but not intros and not patreon stuff). Server not knowing the preference does increase the amount of transmitted data, but maintains user privacy.

> One small issue with this plan is: What should happen if two similar sponsors have different types?

I'm not sure what you mean. Do you mean two similar segments have the "sponsor" type? Then the best-rated one is picked, which is hopefully the most accurate one. If two overlapping segments have two different types, then I would guess only one type is actually correct because the underlying content is the same. Then the best-rated segment that (hopefully) has the correct type is sent.

I would like to see an example, even made up one, of a situation that you are concerned about.

**Comment by ajayyy** [comment]
Now that I think about it, it should be fine. Thanks!

Is this tested, or should I not roll it out to production yet.

**Comment by NilsIrl** [comment]
@bershanskiy

**Comment by ajayyy** [comment]
https://github.com/ajayyy/SponsorBlockServer/pull/71

**Pull request closed without merging**